src/xds/client.rs

```diff
@@ -38,50 +40,64 @@ pub struct ResourceKey {
     pub type_url: String,
 }
 
-#[derive(Default)]
-pub struct HandlerContext {
-    rejects: Vec<RejectedConfig>,
+pub struct RejectedConfig {
+    name: String,
+    reason: anyhow::Error,
 }
 
-impl HandlerContext {
-    pub fn new() -> HandlerContext {
-        HandlerContext {
-            ..Default::default()
-        }
+impl RejectedConfig {
+    pub fn new(name: String, reason: anyhow::Error) -> Self {
+        Self { name, reason }
     }
 }
 
-struct RejectedConfig {
-    name: String,
-    reason: anyhow::Error,
-}
-
 impl fmt::Display for RejectedConfig {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
         write!(f, "{}: {}", self.name, self.reason,)
     }
 }
 
-impl HandlerContext {
-    pub fn reject(&mut self, name: String, reason: anyhow::Error) {
-        self.rejects.push(RejectedConfig { name, reason })
-    }
-}
-
 pub trait Handler<T: prost::Message>: Send + Sync + 'static {
-    fn handle(&self, ctx: &mut HandlerContext, res: Vec<XdsUpdate<T>>);
+    fn handle(&self, res: Vec<XdsUpdate<T>>) -> Result<(), Vec<RejectedConfig>>;
 }
 
 struct NopHandler {}
 
 impl<T: prost::Message> Handler<T> for NopHandler {
-    fn handle(&self, _ctx: &mut HandlerContext, _res: Vec<XdsUpdate<T>>) {}
+    fn handle(&self, _res: Vec<XdsUpdate<T>>) -> Result<(), Vec<RejectedConfig>> {
+        Ok(())
+    }
+}
+
+/// handle_single_resource is a helper to process a set of updates with a closure that processes items one-by-one.
+/// It handles aggregating errors as NACKS.
+pub fn handle_single_resource<T: prost::Message, F: FnMut(XdsUpdate<T>) -> anyhow::Result<()>>(
```

Comment:
Should this logic to do one-by-one processing and aggregating NACKS be moved into AdClient/wherever we call `handle?` 

The two reasons not to would be:
* Hold the same lock for the entire set of updates
* Some kind of interdependence between updates in a set. We'd have to handle cases where updates are not part of the same set, so I don't think this is a valid reason. 

Not sure of a good way to move this with a single lock unless the Handler interface also exposed some `HandleGuard` wrapping the lock but that indirection might be confusing.

Reply:
IMO handlers should be _able_ to operate on the full set, but this helper allows them to effectively opt-in to handling 1-by-1. The motivation was the locking mostly.

Reply:
My point was that operating on the full set is likely not the common case since some set of updates should be considered a small view of a larger continuous stream. 

The only things a handler has to worry about WRT a full set of updates are internal details like locking